utils/waitable_counter.hh

```diff
@@ -0,0 +1,142 @@
+/*
+ * Copyright 2021-present ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <map>
+#include <iosfwd>
+#include <exception>
+
+#include <seastar/core/future.hh>
+#include <seastar/core/future-util.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/shared_future.hh>
+
+namespace utils {
+
+/**
+ * Simple counter type wrapper, that allows
+ * waiting for it to reach a given value.
+ * 
+ * Actual content must be monotonically 
+ * increasing, and unless you want a lot
+ * of entries, waiting should be either
+ * sparse, or at least quantized to "typical"
+ * values (like disk blocks or similar, hint hint)
+ */
+template<typename T, typename Cmp = std::less<T>>
+class waitable_counter {
+private:
+    using shared_promise_type = seastar::shared_promise<>;
+
+    T _value;
+    std::map<T, shared_promise_type, Cmp> _waiters;
```

Comment:
Perhaps we can prematurely optimize into a abseil btree map, and remove an allocation.

Note there's no need for shared_promise, shared_future is enough (it acts as a factory for futures).

Reply:
And maybe circular_buffer<entry> is better, do we need random access?

Reply:
Simplicity. std::map is nicely ordered, and neat to access. Can use intrusive_map or similar just as well. Or even self-sort a vector. Again, keeping it simple until proven it needs not be.

Reply:
What's the expected element count?

Reply:
Current usage, 1-10 elements. Again, my testing probably does not stress even close to enough to really say, but I have max 2-3 waiters.